probe-rs/src/probe/flash/builder.rs

```diff
@@ -303,115 +301,100 @@ impl<'a> FlashBuilder<'a> {
-                            added_new_sector = true;
-                            log::debug!("Adding Sector");
+                            log::debug!("Added Sector");
                         } else {
                             return Err(FlashBuilderError::InvalidFlashAddress(flash_address));
                         }
+                        continue;
+                    } else {
+                        if let Some(page) = sector.pages.last_mut() {
+                            // If the current page does not contain the address.
+                            dbg!(flash_address);
+                            dbg!(page.address, page.size);
+                            if flash_address >= page.address + page.size {
+                                // Fill any gap at the end of the current page before switching to a new page.
+                                Self::fill_end_of_page_gap(
+                                    flash,
+                                    page,
+                                    page.size as usize - page.data.len(),
+                                    keep_unwritten,
+                                )?;
+
+                                let page_info = flash.region().get_page_info(flash_address);
+                                if let Some(page_info) = page_info {
+                                    let new_page = FlashPage::new(&page_info);
+                                    dbg!(sector.pages.len());
+                                    dbg!(&new_page);
+                                    sector.add_page(new_page)?;
+                                    log::debug!("Added Page");
+                                } else {
+                                    return Err(FlashBuilderError::InvalidFlashAddress(flash_address));
+                                }
+                                continue;
+                            } else {
+                                let space_left_in_page = page.size - page.data.len() as u32;
+                                let space_left_in_data = op.data.len() - pos;
+                                let amount = usize::min(space_left_in_page as usize, space_left_in_data);
+
+                                page
+                                    .data
+                                    .extend(&op.data[pos..pos + amount]);
+                                log::debug!("Added {} bytes to page", amount);
+                                pos += amount;
+                            }
+                        } else {
+                            // If no page is on the sector yet.
+                            let page_info = flash.region().get_page_info(flash_address);
+                            if let Some(page_info) = page_info {
+                                let new_page = FlashPage::new(&page_info);
+                                sector.add_page(new_page)?;
+                                log::debug!("Added Page");
```

Comment:
This doesn't sound terribly helpful. Maybe also log the page? And this should probably be `trace` as it happens very often when flashing something large.